### PR TITLE
graphene-hardened-malloc: 2 -> 8, overhaul tests

### PR DESCRIPTION
**Pull request by risicle**
###### Motivation for this change
Supersedes #115546

Upstream's tests now have a real driver, so overhaul our use of them to cover both build-time-linking and `LD_PRELOAD` use, and simplifying the hardened nixos test to allow it to reuse this test setup.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc @NeQuissimus @zzywysm

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 133556` at 5a9bfc5c run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>graphene-hardened-malloc</li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `writeScript`.
Near pkgs/development/libraries/graphene-hardened-malloc/default.nix:1:35:
```
  |
1 | { lib, stdenv, fetchurl, python3, writeScript, runCommand }:
  |                                   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`checkPhase` should probably contain `runHook preCheck` and `runHook postCheck`.

Near pkgs/development/libraries/graphene-hardened-malloc/default.nix:16:3:
```
   |
16 |   checkPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/development/libraries/graphene-hardened-malloc/default.nix:20:3:
```
   |
20 |   installPhase = ''
   |   ^
```
</li>
  </ul>
</details>

**Review by Artturin — APPROVED** [review]
tests run

**Comment by risicle** [comment]
@ofborg test hardened

**Comment by risicle** [comment]
These tests cover failure modes pretty well now, but something it doesn't really test much is the _correct_ operation of it as a `malloc`. I wonder if I should consider adding some sort of malloc torture test. Any suggestions?

**Comment by risicle** [comment]
Oh I've also noticed that my method of running the tests from `nixosTests.hardened` slightly defeats the point, because it wraps them in a `preload-hardened-malloc` call meaning we're not _actually_ testing the `environment.memoryAllocator.provider` setting's correct operation.

**Comment by risicle** [comment]
@ofborg test hardened

**Comment by Artturin** [comment]
hardened test runs

**Comment by risicle** [comment]
I'm just going to give a hardened maintainer a chance to weigh in before I merge...

File: pkgs/development/libraries/graphene-hardened-malloc/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/graphene-hardened-malloc/default.nix:13]:
```suggestion
  };

  doCheck = true;
```

File: pkgs/development/libraries/graphene-hardened-malloc/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/graphene-hardened-malloc/default.nix:42]: This is missing -j$NIX_BUILD_CORES

File: pkgs/development/libraries/graphene-hardened-malloc/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/graphene-hardened-malloc/default.nix:57]: Why not use makeWrapper?

Comment (risicle) [reply pkgs/development/libraries/graphene-hardened-malloc/default.nix:57]: Yeah guess I can now I've got rid of the fancy nonsense..

File: pkgs/development/libraries/graphene-hardened-malloc/default.nix
Comment (risicle) [reply pkgs/development/libraries/graphene-hardened-malloc/default.nix:43]: I'd rather this take 1 second instead of 0.5 seconds and not have to break the already-long line.

**Comment by risicle** [comment]
@ofborg test hardened

**Pull request merged**
